text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,218 @@
+4. if the registry gave you a `key-subject` then on the command line run `cargo login --registry=name --key-subject="the provided data"`
+
+There are credential processes for using key pairs stored on hardware tokens. Check crates.io to see if there's one available for your hardware. Each one is a little different, but the general workflow is:
+1. `cargo install credential-process-for-your-hardware-token`
+2. Run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
+3. Edit `credentials.toml` to have a `credential-process` field as described by `credential-process-for-your-hardware-token` docs. (The credential process command may help do this for you.)
+4. Log into the registries website
+5. Go to the "register a key pair" page, upload your public key.
+
+Some registries prioritize user experience over strictest security. They can simplify the process by providing key generation on the server. If your registry works this way the workflow will be:
+1. Log into the registries website
+2. Go to the "register generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not keep a copy of the public key!
+3. Run it on the command line. It will look like  `cargo login --registry=name --private-key="key"`
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In [`config.toml`](https://doc.rust-lang.org/cargo/reference/config.html) and `credentials.toml` files there is a field called `private-key`, witch is a private key formatted in the secret [subset of `PASERK`](https://github.com/paseto-standard/paserk/blob/master/types/secret.md) and is used to sign asymmetric tokens
+
+A keypair can be generated with `cargo login --generate-keypair` which will:
+- generate a public/private keypair in the currently recommended fashion.
+- save the private key in `credentials.toml`.
+- print the public key and the path to the file.
+  (See unresolved questions section.)
+
+It is recommended that the `private-key` be saved in `credentials.toml`. It is also supported in `config.toml`, primarily so that it can be set using the associated environment variable. Witch is the recommended way to provide it in CI contexts. This set up is what we have for the `token` field for setting a secret token.
+
+There is also an optional field called `private-key-subject` which is a string chosen by the registry.
+This string will be included as part of an asymmetric token and should not be secret.
+It is intended for the rare use cases like "cryptographic proof that the central CA server authorized this action". Cargo requires it to be non-whitespace printable ASCII. Registries that need non-ASCII data should base64 encode it.
+
+Both fields can be set with `cargo login --registry=name --private-key="key" --private-key-subject="subject"`.
+
+A registry can have at most one of `private-key`, `token`, or `credential-process` set.
```

Comment:
Should the private key have some kind of prefix so we can identify it as PASETO v3 in case we ever want to change to a different key format?

Reply:
v3 Private keys encoded in PASERK secret format start with `k3.secret.`. So I think it will be easy to tell that the `private-key` was intended for this RFC.